### ISSUE TITLE
Flexible and efficient optionally-structured console logging out of the box

### ISSUE THREAD

**Issue by jskeet** [aspnet/AspNetCore#12237]
### Is your feature request related to a problem? Please describe.

As a web app developer, I'm trying to write web apps (such as nodatime.org) that log efficiently and cleanly, with minimal effort.
As a Google Cloud Platform client library author, I'm trying to make it really easy for our users to integrate with Stackdriver logging and error reporting.
As a .NET community member, I'm trying to make [12-factor apps](https://12factor.net/) easier to write.

Currently, there are three reasonably obvious options for logging in my web application:

**Option 1: Use the default console logger.**

The "no code changed" option.

This has at least three problems:

- Each log entry spans two lines, which makes it far less useful than it might be, and presumably means log entries on busy systems could be interleaved.
- The log entries contain control characters for colour; I'm sure I can disable this, but it's annoying out-of-the-box behaviour
- It's not structured in an easily machine-readable way.

**Option 2: Use Google.Cloud.Diagnostics.AspNetCore**

This library makes RPCs directly to the Stackdriver Logging API.

As one of the contributors to the library, I should be able to get this to work easily, right? Well, it does work, but not as conveniently as I'd like. I can improve things, but this is likely to always be a somewhat-heavyweight option. Useful in some cases, such as when there's nothing monitoring console output, but not necessary in modern container-based systems which almost always *do* monitor the console. It also requires appropriate credentials to be provided when not running on GCP, etc. This ties it to Stackdriver in a way that could be harmful to app portability. (It's fine if you don't need portability of course, but it's a heavyweight dependency otherwise.)

**Option 3: Use Serilog**

I did get this working, writing out structured JSON to the console, but it was relatively painful to do, and didn't write the JSON in a format that Stackdriver expects. In order to get a custom JSON format, I had to do far more work than I'd normally expect to. It worked, but there's no way that every app should have all of this code in it.

### Describe the solution you'd like

Ideally I'd like three simple options:;

- A more concise single-line-per-entry console logger, still just lines of plain text
- A simple structured logger writing JSON in whatever format seems most reasonable after some research
- A way of adapting the second option to write in a custom JSON format

The last is for integration with Stackdriver and presumably other cloud providers' monitoring offerings. With the right JSON, the Stackdriver user interface is clear and easy to read, but still provides structured data that can be used to trigger error reporting, create filters etc.

In an ideal world the CNCF would provide a common JSON format for everyone to use for maximum interoperability, but until that happens it would be useful for the ASP.NET Core team to handle the concerns of gathering the log data and writing the resulting to the console efficiently and atomically, leaving cloud providers to just provide the code to convert the raw log data into appropriate JSON. Even though an app using this would still have a dependency on Stackdriver, it's a much lighter-weight one than Google.Cloud.Diagnostics.AspNetCore - and it would be entirely reasonable to depend on multiple formatters and pick the right one at execution time, expecting each to be standalone and small.

### Additional context

I have *tried* to find existing solutions to this, but I've failed so far.

For nodatime.org, I ended up writing my own console logger, which should only be used for the sake of seeing what I'm trying to achieve - it's not aimed at high-performance scenarios, for example. (If anyone wants to point and laugh, they're welcome - it does the job though :)
https://github.com/nodatime/nodatime.org/blob/master/src/NodaTime.Web/Logging/JsonConsoleLogger.cs

**Comment by secana** [comment]
We run into the sample problems frequently. The default logger is not usable for log aggregation in a service oriented architecture. Structured logging, like Serilog provides it, would be a much better default. Logs with color and two lines per message break every log aggregations system and renders them useless.
This means that we have to provide our own Logger assembly in every service which uses Serilog for the ILogger interface. An easy to use default implementation with structured logging, directly provided through Microsoft.Extensions.Logging would be much preferred.

**Comment by paulomorgado** [comment]
Microsoft.Extensions.Logging provides structured logging. You wouldn't be able to use Serilog as a provider if it didn't.

**Comment by jskeet** [comment]
@paulomorgado: Yes, it's definitely possible to use structured logging with ASP.NET Core - it's just that it's relatively painful to do out-of-the-box, especially in a 12-factor friendly (stdout) way. With a more flexible console logger, it would be really simple.

**Comment by paulomorgado** [comment]
I'm sorry, @jskeet, but from what I've seen in the wild, I would make console logging physically painful for production logging. 😄

**Comment by jskeet** [comment]
@paulomorgado: Whereas in the wild I've found that with things like Stackdriver logging + fluentd automatically capturing stdout, it's a *lot* simpler to handle logging just to the console than having to configure logging in more complex ways. Each to their own, but I'd at least like it to be a reasonable option. You can still avoid using the console if you want to :)

**Comment by secana** [comment]
@paulomorgado Logging to stdout is the default for distributed micro-services with Docker. E.g. Docker Swarm/K8s greps all stdout and forwards it to Logstash/fluentd or similar. It should be as easy as possible not as painful as possible for production.

**Comment by paulomorgado** [comment]
@secana, how do you go from screen grepping to structure logging?

**Comment by jskeet** [comment]
@paulomorgado: In the case of Stackdriver at least, if you write a line of JSON to the console, that's parsed and treated in a structured way. I'd certainly *hope* that other log collectors work that way too.

**Comment by secana** [comment]
@jskeet yes, that is how it works for all other collectors that I know of.

**Comment by nblumhardt** [comment]
@jskeet slightly O/T, but would appreciate your feedback:

> In order to get a custom JSON format, I had to do far more work than I'd normally expect to.

seems at odds with:

> A simple structured logger writing JSON in whatever format seems most reasonable after some research

Did you find _Serilog.Formatting.Compact_? It's an implementation of the latter (opinionated/reasonable JSON rendering). Thanks! :-)

**Comment by jskeet** [comment]
@nblumhart: I don't see how those two statements are at odds at all. I'd like there to be an out-of-the-box solution for a simple structured JSON logger that writes in some reasonable format... if there's something that multiple log consumers will accept, that would be great. But it still needs to be adaptable as per the bullet that follows.

Yes, I found Serilog.Formatting.Compact. It did indeed write out JSON to the console - after more hoops than I'd like to see in an out-of-the-box ASP.NET Core implementation - but it did so in its own unconfigurable format, which Stackdriver can't do much with. Stackdriver expects specific JSON properties, specific severity levels etc. [CompactJsonFormatter](https://github.com/serilog/serilog-formatting-compact/blob/master/src/Serilog.Formatting.Compact/Formatting/Compact/CompactJsonFormatter.cs
) is completely hard-coded in that respect, which means I can't use it with Stackdriver. (It also doesn't format the message itself, which makes perfect sense for a *compact* formatter, but also makes it less useful for Stackdriver.)

**Comment by nblumhardt** [comment]
Thanks for the reply, @jonskeet!

I initially misread your third point- 

> A way of adapting the second option to write in a custom JSON format

Sorry about my confusion there, on re-reading no I understand what you meant :+1:

My interest in "configurable vs. opinionated" comes from having started with a configurable `JsonFormatter` in Serilog, but later backpedalling to the much simpler and more efficient `CompactJsonFormatter` /`RenderedCompactJsonFormatter` after failing to put limits on the flexibility of the original implementation. Adding configurability in there without sacrificing perf, usability, or maintainability is a challenge, and we didn't want to end up with essentially a clone of JSON.NET :-)

Will be interesting to see what kinds of approaches are possible, here.

Just RE jumping through hoops, adding a NuGet package and `WriteTo.Console(new CompactJsonFormatter())` seems like a reasonable API/experience, anything we should change or improve? Cheers!

**Comment by nblumhardt** [comment]
Ah, also - sorry

> (It also doesn't format the message itself, which makes perfect sense for a compact formatter, but also makes it less useful for Stackdriver.)

The example in the README shows the `CompactJsonFormatter` variant, but [`RenderedCompactJsonFormatter`](https://github.com/serilog/serilog-formatting-compact/blob/dev/src/Serilog.Formatting.Compact/Formatting/Compact/RenderedCompactJsonFormatter.cs) is probably much closer to what you need, and does format the message in there.

**Comment by jskeet** [comment]
@nblumhardt: This is the code I ended up with in my Program.cs:

```csharp
public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
    WebHost.CreateDefaultBuilder(args)
        .UseSerilog((hostingContext, loggerConfiguration) => loggerConfiguration
            .ReadFrom.Configuration(hostingContext.Configuration)
            .Enrich.FromLogContext()
            .WriteTo.Console(new StackdriverJsonFormatter()))
        .UseStartup<Startup>();
```

For someone who hasn't used Serilog before, that took a while to get right (including finding all the right packages to start with - there was a fair amount of trial and error and trying to find docs, checking which packages I'd need etc) and seems unnecessarily verbose. I'd expect *most* ASP.NET Core Serilog users to want to read from the regular ASP.NET Core configuration and enrich from the log context. If we could just have:

```csharp
public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
    WebHost.CreateDefaultBuilder(args)
        .UseSerilog(sinkConfig => sinkConfig.WriteTo.Console(new StackdriverJsonFormatter()))
        .UseStartup<Startup>();
```

that would significantly lower the barriers to entry for someone who doesn't need or care about most of the Serilog features, but just wants to use a specific log renderer. However, this issue is intended to be about ASP.NET Core built-in usability rather than Serilog - if you'd like me to file a friction log with my experiences, I'm happy to do that (although I'd be starting again; I don't think I kept a detailed friction log of that at the time). I don't think users should have to use Serilog just to write a reasonable 12factor app with ASP.NET Core.

**Comment by nblumhardt** [comment]
Thanks for the feedback! Any further feedback you can send our way would be much appreciated. We're edging towards a more "batteries included" default configuration in _Serilog.AspNetCore_, hooking into configuration and the log context by default definitely seems worth us considering, there 👍 

Let me know if you need another set of eyes on `StackdriverJsonFormatter`, happy to dig through and see if there are any APIs/goodies in the Serilog ecosystem that might help with it.
